cpp/src/arrow/adapters/orc/adapter.cc

```diff
@@ -183,6 +183,22 @@ liborc::RowReaderOptions default_row_reader_options() {
   return options;
 }
 
+// Proactively check the availability of timezone database.
+// Remove it once https://issues.apache.org/jira/browse/ORC-1661 has been fixed.
+Status check_timezone_database_availability() {
```

Comment:
Is this useful if we're correctly catching exceptions? I'm afraid this could get out of sync with ORC's own timezone loading coding.

I'd rather not do this, and let ORC fix the issue by making the timezone file optional (what is it used for exactly?).

Reply:
@wgtmac what would be the error message that you get when not doing the above now that you are catching the exceptions?

Because one reason to keep this, is to provide an informative error message (most Windows users will see this the first time, so I think it is important to give some guidance in the error message, and not just a "file not found"). 
(to avoid getting out of sync with ORC, we could add an ORC version check and only do this for ORC<2?)

Reply:
After catching all exceptions, the error message here without the `check_timezone_database_availability()` is `Invalid: Can't open /usr/share/zoneinfo/XXX`.

Reply:
>  (to avoid getting out of sync with ORC, we could add an ORC version check and only do this for ORC<2?)

That sounds reasonable to me.

Reply:
> I'd rather not do this, and let ORC fix the issue by making the timezone file optional (what is it used for exactly?).

ORC has two timestamp types: timestamp (namely timestamp_without_timezone) and timestamp_instant (namely timestamp_with_local_timezone). 
- timestamp: writer keeps the writer timezone in the stripe footer, and reader uses reader timezone to recover the same wall clock time by converting from writer timezone to reader timezone. that's why the reader try to call `getLocalTimezone()` on startup.
- timestamp_with_local_timezone: use UTC everywhere so it does not have any conversion.

Reply:
> * timestamp: writer keeps the writer timezone in the stripe footer, and reader uses reader timezone to recover the same wall clock time by converting from writer timezone to reader timezone. that's why the reader try to call `getLocalTimezone()` on startup.

If we convert, we should convert to UTC instead of converting to the local timezone. That's how Arrow timestamps work (but we still need a timezone DB for the conversion anyway):
https://github.com/apache/arrow/blob/51817917e1436c8799ed382d160798060cd76652/format/Schema.fbs#L283-L288

Reply:
> If we convert, we should convert to UTC instead of converting to the local timezone.

I know that's a better idea. However, this was the design choice of ORC timestamp type which originates from Hive. So it would be better to use timestamp_instant type in favor of the old timestamp type in ORC at any time.